components/form.js

```diff
@@ -0,0 +1,178 @@
+        }
+        return isValid;
+    }
+
+    getToken() {
+        let token = JSON.parse(localStorage.getItem('Bearer Token'));
+        if (!token || this.isTokenValid()) {
+            fetch(baseUrl + '/oauth/token', {
+                method: 'POST',
+                body: JSON.stringify({
+                    scope: '*',
+                    client_secret: '35e7f0bca957836d05ca0492211b0ac707671261',
+                    client_id: 'ushahidiui',
+                    grant_type: 'password',
+                    password: Person.password,
+                    username: Person.userName
+                }),
+                headers: {
+                    Accept: 'application/json',
+                    'Content-Type': 'application/json',
+                    Authorization: 'large-auth-string'
+                }
+            })
+                .then(response => {
+                    if (response.ok) {
+                        return response.json();
+                    } else {
+                        throw new Error('Token fetching response is not ok.');
+                    }
+                })
+                .then(data =>
+                    this.setState({
+                        token: { ...data, fetching_time: new Date().getTime() }
+                    })
+                )
+                .catch(error => console.log(error));
+        } else {
+            this.setState({ token: token });
+        }
+    }
+
+    getFormFields() {
+        // Fetch the data form fields from API, use it in state
+        let formFields = JSON.parse(localStorage.getItem('Form Fields'));
+        if (!(Array.isArray(formFields) && formFields.length)) {
+            fetch(
+                baseUrl +
+                    '/api/v3/forms/14/attributes?order=asc&orderby=priority'
+            )
+                .then(response => {
+                    if (response.ok) {
+                        return response.json();
+                    } else {
+                        throw new Error('Field fetching response is not ok.');
+                    }
+                })
+                .then(data =>
+                    this.setState({
+                        formFields: data.results
+                    })
+                )
+                .catch(err => console.log(err));
+        } else {
+            this.setState({
+                formFields: formFields
+            });
+        }
+    }
+
+    componentDidMount() {
+        this.getFormFields();
+        this.getToken();
+    }
+
+    componentDidUpdate(prevProps, prevState) {
+        localStorage.setItem(
+            'Form Fields',
+            JSON.stringify(this.state.formFields)
+        );
+        localStorage.setItem('Bearer Token', JSON.stringify(this.state.token));
+    }
+
+    // Handler html For when the submit button is pressed
+    onSubmit = e => {
+        e.preventDefault();
+
+        // Make a post request to Ushahidi sever
+        fetch(baseUrl + '/api/v3/posts', {
```

Comment:
this and other fetch() calls would probably be a good candidate to move them to a service in another js file

Reply:
done